daffodil-io/src/main/scala/org/apache/daffodil/io/DirectOrBufferedDataOutputStream.scala

```diff
@@ -33,30 +33,79 @@ import org.apache.daffodil.util.Maybe
 import org.apache.daffodil.util.Maybe._
 import org.apache.daffodil.util.MaybeULong
 import org.apache.daffodil.util.Misc
+import java.io.File
 
 /**
  * This simple extension just gives us a public method for access to the underlying byte array.
  * That way we don't have to make a copy just to access the bytes.
  */
-private[io] class ByteArrayOutputStreamWithGetBuf() extends java.io.ByteArrayOutputStream {
+private[io] class ByteArrayOrFileOutputStreamWithGetBuf(
+  chunkSizeInBytes: Long = 4096,
+  maxBufferSizeInBytes: Long = 2000 * (1 << 20))
+  extends java.io.ByteArrayOutputStream {
+  private var _byteArrayOS = this
+  private var _fileOS: java.io.FileOutputStream = null
+
+  var isFile: Boolean = false
+  var tempFile: File = null
+  var bytesWritten = count
+
+  def switchOS() = {
+    Assert.invariant(_fileOS == null)
+    tempFile = File.createTempFile("daffodil-", ".tmp")
+    _fileOS = new java.io.FileOutputStream(tempFile)
+    this.flush
+    this.writeTo(_fileOS)
+    bytesWritten = count
+    this.reset
+    isFile = true
+  }
+
+  override def write(b: Array[Byte]) = {
+    if ((_fileOS == null) && ((count + b.length) > maxBufferSizeInBytes))
+      switchOS()
+
+    if (_fileOS != null) {
+      _fileOS.write(b)
+      bytesWritten += b.length
+    } else
+      super.write(b)
+  }
+
+  override def write(b: Array[Byte], off: Int, len: Int) = {
+    if ((_fileOS == null) && ((count + len) > maxBufferSizeInBytes))
+      switchOS()
+
+    if (_fileOS != null) {
+      _fileOS.write(b, off, len)
+      bytesWritten += len
+    } else
+      super.write(b, off, len)
+  }
+
+  override def write(b: Int) {
+    if ((_fileOS == null) && ((count + 1) > maxBufferSizeInBytes))
+      switchOS()
+
+    if (_fileOS != null) {
+      _fileOS.write(b)
+      bytesWritten += 1
+    } else
+      super.write(b)
+  }
+
+  override def close() = if (_fileOS != null) _fileOS.close() else super.close()
+  override def flush() = if (_fileOS != null) _fileOS.flush() else super.flush()
+
+  def getCount() = if (isFile) bytesWritten else count
   def getBuf() = buf
```

Comment:
Should getBuf fail if you have switched to a file output stream? Assert.usage(_fileOS eq null, ...)